Implement A-Star

Kai-Shafe opened the pull request:
Bugfixes:

- [ ] FindNearestBoundaryPoint() now properly maps outside coordinates to the proper boundary point, and now handles when the goal coordinate is inside the maximum boundary. 
- [ ] FindNearestBoundaryPoint() and RoundUTMCoordinate() now return a new UTMCoordinate struct to avoid mutating the original coordinate.
- [ ] ValidCoordinate() boundary check statement logic was incorrect, it now properly includes internal coordinates and excludes external coordinates.
- [ ] AStarNodes now contain a shared_ptr to the parent instead of a raw pointer. This allows for heap allocation and garbage collection of the node, and will guarantee pointer validity for parent nodes when dereferenced in ConstructPath(). This change is needed since the underlying memory of vClosedList is not guaranteed to be static, so I wasn't able to initialize pointers to the correct nodes and have them remain valid. This was the solution that I found this to be the simplest.
- [ ] Changed goal evaluation in PlanAvoidanceRoute() to use GeoMeasurements instead of directly comparing UTMCoordinates.


Updates:

- [ ] AvoidanceState now handles when PlanAvoidanceRoute() fails to generate a path by handling eEndObstacleAvoidance.
- [ ] Added Sharpen kernel to PreprocessFrame() in TagDetector.cpp since this change wasn't published to the Aruco branch properly.
- [ ] Changed AvoidanceState to use the TimeIntervalStuckDetector struct.
- [ ] ConstructPath() no longer uses std::reverse(), now the function uses the recursion structure of the function to make sure the path coordinate vector are placed in the correct order.
- [ ] Changed CalculateNodeHValue() to use GeoMeasurements instead of diagonal distance (This change can be reverted if the performance of PlanAvoidanceRoute() needs improvement). 
- [ ] Removed outdated Python codebase variables in AvoidanceState.

ClayJay3 commented on src/states/AvoidanceState.cpp:
```diff
@@ -70,23 +69,12 @@ namespace statemachine
-        // TODO: Clear ASTAR Pathfinder
-        // TODO: Clear Stanley Controller
+        // Clear Stanley Controller
+        std::vector<geoops::UTMCoordinate> v_clear = {};
```
This seems kinda hacky to me. Add a ClearPath() or ResetController() method to the StanleyController class.

ClayJay3 commented on src/states/AvoidanceState.cpp:
```diff
@@ -117,8 +109,46 @@ namespace statemachine
     {
-        // TODO: Implement the behavior specific to the Avoidance state
         LOG_DEBUG(logging::g_qSharedLogger, "AvoidanceState: Running state-specific behavior.");
+
+        // TODO: build in obstacle detection and refresh of astar path when a new object is detected
+        // This can be done by calling PlanAvoidanceRoute() with an updated obstacle vector.
+
+        // A route has already been plotted by the planner and passed to the controller.
+        // Navigate by issuing drive commands from the controller.
+        // Check if we are at the goal waypoint.
+        geoops::RoverPose stCurrentPose                  = globals::g_pWaypointHandler->SmartRetrieveRoverPose();
```
From what I can tell, this currently doesn't follow the generated path by using astar and the stanleycontroller. Currently works that same as the navigating state.

Kai-Shafe replied:
Yeah... something I missed that might invalidate a lot of my testing. I've changed AvoidanceState to call the StanleyController::Calculate() method.

ClayJay3 commented on src/states/AvoidanceState.cpp:
```diff
@@ -117,8 +109,46 @@ namespace statemachine
+        {
+            globals::g_pDriveBoard->SendStop();
+            globals::g_pStateMachineHandler->HandleEvent(Event::eEndObstacleAvoidance, false);
+        }
```
This comment refers to the HandleEvent method.

```
case Event::eEndObstacleAvoidance:
    {
        LOG_INFO(logging::g_qSharedLogger, "AvoidanceState: Handling EndObstacleAvoidance event.");
        eNextState = States::NUM_STATES;    // Replace with `get_prev_state()`
        break;
    }
```

If this event ever gets triggered in its current code state, then the statemachine will be stuck in avoidance since the next state doesn't exist. Look at the stuck state if you want an example that shows how to return to the state that triggered avoidance.

ClayJay3 commented on src/states/AvoidanceState.h:
```diff
@@ -31,21 +35,16 @@ namespace statemachine
+            statemachine::TimeIntervalBasedStuckDetector m_stStuckChecker;
+
+            pathplanners::AStar m_stPlanner;
+            controllers::StanleyController m_stController =
```
Do not instantiate class member objects in the header file, move to constructor.

ClayJay3 commented on src/util/planners/Nodes.hpp:
```diff
@@ -57,11 +57,11 @@ namespace pathplanners
-                double dKg;                              // The cost of the path from the star node to the current node.
-                double dKh;                              // The heuristic estimate of how much it will cost to get to the end node.
-                double dKf;                              // The sum of dKg and dKh. An estimate for the total cost of the node.
+                std::shared_ptr<AStarNode> stdParentNode;    // A shared pointer to the parent node the this nodes comes after in the path.
```
Change name to pParentNode since this is still a pointer.

ClayJay3 commented on src/vision/aruco/ArucoDetection.hpp:
```diff
@@ -97,6 +97,7 @@ namespace arucotag
     {
         // Grayscale.
         cv::cvtColor(cvInputFrame, cvOutputFrame, cv::COLOR_BGRA2GRAY);
+        cv::filter2D(cvInputFrame, cvInputFrame, -1, constants::ARUCO_SHARPEN_KERNEL_EXTRA);
```
What is the performance like on this? Just curious.

ClayJay3 replied:
Also change the `cvInputFrame` to `cvOutputFrame`. 

`cv::filter2D(cvOutputFrame, cvOutputFrame, -1, constants::ARUCO_SHARPEN_KERNEL_EXTRA);`

Kai-Shafe replied:
It's been a while since I had tested it (and I'm not really sure how to test it now with my current setup lol), but from what I remember the ..._KERNEL_FAST dropped us to around 20-25 FPS, and the ..._KERNEL_EXTRA dropped us to about 15-20 FPS on a camera that was getting 30 FPS.

Byrdman32 commented on tests/Unit/src/algorithms/pathplanners/AStarPlanner.cc:
```diff
@@ -0,0 +1,261 @@
+ * @author ClayJay3 (claytonraycowen@gmail.com)
+ * @date 2023-10-12
+ *
+ * @copyright Copyright MRDT 2023 - All Rights Reserved
```
Should be copyright 2024:

` * @copyright Copyright MRDT 2024 - All Rights Reserved`

Byrdman32 commented on tests/Unit/src/algorithms/pathplanners/AStarPlanner.cc:
```diff
@@ -0,0 +1,261 @@
+/******************************************************************************
+ * @brief Unit test for PIDController algorithm class.
```
Change brief to:

` * @brief Unit test for AStar algorithm class.`

Byrdman32 commented on tests/Unit/src/algorithms/pathplanners/AStarPlanner.cc:
```diff
@@ -0,0 +1,261 @@
+ * @brief Unit test for PIDController algorithm class.
+ *
+ * @file PIDController.cc
+ * @author ClayJay3 (claytonraycowen@gmail.com)
```
Change to your name and email:

` * @author Kai Shafe (kasq5m@umsystem.edu)`

Byrdman32 commented on tests/Unit/src/algorithms/pathplanners/AStarPlanner.cc:
```diff
@@ -0,0 +1,261 @@
+ *
+ * @file PIDController.cc
+ * @author ClayJay3 (claytonraycowen@gmail.com)
+ * @date 2023-10-12
```
Change to today's date:

` * @date 2024-04-28`

Byrdman32 commented on tests/Unit/src/algorithms/pathplanners/AStarPlanner.cc:
```diff
@@ -0,0 +1,261 @@
+/******************************************************************************
+ * @brief Unit test for PIDController algorithm class.
+ *
+ * @file PIDController.cc
```
Change to correct file name:

` * @file AStarPlanner.cc`

Byrdman32 commented on tests/Unit/src/algorithms/pathplanners/AStarPlanner.cc:
```diff
@@ -0,0 +1,261 @@
+    delete pAStar;
+    // Point to null.
+    pAStar = nullptr;
+}
```
Add blank line to end of file.

Byrdman32 reviewed (changes requested):
Same things that Clayton noted. Also some notes on the Unit Test.

Kai-Shafe commented on src/algorithms/planners/AStar.cpp:
```diff
@@ -361,15 +367,17 @@ namespace pathplanners
         }
         // Recursive case: call ConstructPath on parent pointer.
-        ConstructPath(*stEndNode.stParentNode);
+        ConstructPath(*stEndNode.stdParentNode);
```
When I was running through the code with the debugger, this was one of the methods I thoroughly traced as it was the primary one giving me issues. When I traced it with the current implementation, it did actually dereference stdParentNode instead of stEndNode, and correctly built the path while on the sim. When I tried adding the additional dereference you mentioned I got an error.

Kai-Shafe commented on src/algorithms/planners/AStar.cpp:
```diff
@@ -361,15 +367,17 @@ namespace pathplanners
-        m_vPathCoordinates.emplace_back(stEndNode.stNodeLocation);
         // Base case: Check for origin node.
-        if (stEndNode.stParentNode == nullptr)
+        if (stEndNode.stdParentNode == nullptr)
```
When I checked CONTRIBUTING.md, that was the stated prefix for std types, which is where I got the shared_ptr object. Originally stParentNode was misnamed as it was a pointer, and the latest push has changed the prefix to be `pParentNode`.

Kai-Shafe commented on src/algorithms/planners/AStar.cpp:
```diff
@@ -70,8 +70,6 @@ namespace pathplanners
      * @brief Helper function to destroy objects from m_vObstacles.
      *
-     * @todo Determine if this needs to be public for error handling.
-     *
```
UpdateObstacleData() now has a second bool param (default: true) named `bClearObstacles` that will optionally clear the obstacles vector within AStar.

ClayJay3 reviewed (commented):
Looks good, functionality need to be checked via the unit test which currently segfaults. Then PR can be merged.

Kai-Shafe commented:
During debugging the unit test for `PlanAvoidanceRoute()` I discovered that my code would have a critical failure if the start or end `UTMCoordinate` structs were not initialized fully, which would result in a `GeoMeasurement` struct being returned with uninitialized data. I added a procedure for checking for uninitialized data, utilizing Euclidian distance on just the Easting and Northing values of two coordinates. As of right now, my code now properly returns a valid path during the unit tests.

I was wondering if there was any better way to initialize `UTMCoordinate` structs in the unit tests so I can avoid the error being logged on line 443 in GeospatialOperations.hpp? I feel as though the handling I added was a bit hack-y and isn't ideal. Any advice is appreciated!

Byrdman32 commented:
> During debugging the unit test for `PlanAvoidanceRoute()` I discovered that my code would have a critical failure if the start or end `UTMCoordinate` structs were not initialized fully, which would result in a `GeoMeasurement` struct being returned with uninitialized data. I added a procedure for checking for uninitialized data, and handled it by utilizing Euclidian distance on just the Easting and Northing values of two coordinates. As of right now, my code now properly returns a valid path during the unit tests.
> 
> I was wondering if there was any better way to initialize `UTMCoordinate` structs in the unit tests so I can avoid the error being logged on line 443 in GeospatialOperations.hpp? I feel as though the handling I added was a bit hack-y and isn't ideal. Any advice is appreciated!

I'll look at this in the morning.

Byrdman32 commented on tests/Unit/src/algorithms/pathplanners/AStarPlanner.cc:
```diff
@@ -0,0 +1,339 @@
+ * @author Kai Shafe (kasq5m@umsystem.edu)
+ * @date 2024-4-28
+ *
+ * @copyright Copyright MRDT 2024 - All Rights Reserved
```
Change to `* @copyright Copyright Mars Rover Design Team 2024 - All Rights Reserved`

Byrdman32 reviewed (approved):
Looks really good. Please make the minor change to the doxygen header in `AStarPlanner.cc` and it looks good to me.

ClayJay3 commented on src/drivers/DriveBoard.cpp:
```diff
@@ -121,8 +121,8 @@ void DriveBoard::SendDrive(diffdrive::DrivePowers& stDrivePowers)
-    float fDriveBoardLeftPower  = numops::MapRange(float(dLeftSpeed), -1.0f, 1.0f, m_fMinDriveEffort, m_fMaxDriveEffort);
-    float fDriveBoardRightPower = numops::MapRange(float(dRightSpeed), -1.0f, 1.0f, m_fMinDriveEffort, m_fMaxDriveEffort);
+    float fDriveBoardLeftPower  = numops::MapRange(float(dLeftSpeed), -1.0F, 1.0F, m_fMinDriveEffort, m_fMaxDriveEffort);
+    float fDriveBoardRightPower = numops::MapRange(float(dRightSpeed), -1.0F, 1.0F, m_fMinDriveEffort, m_fMaxDriveEffort);
```
Change to lower.

ClayJay3 commented on src/vision/aruco/TagDetector.h:
```diff
@@ -78,7 +78,7 @@ class TagDetector : public AutonomyThread<void>
         /////////////////////////////////////////
 
-        void EnableTensorflowDetection(const float fMinObjectConfidence = 0.4f, const float fNMSThreshold = 0.6f);
+        void EnableTensorflowDetection(const float fMinObjectConfidence = 0.4F, const float fNMSThreshold = 0.6F);
```
Change to lower

ClayJay3 reviewed (approved):
Nice tests.
